Add tiered caching blog (#3376)
opensearch-project/project-website: _posts/2024-10-11-tiered-cache.md

---
layout: post
title:  "Tiered caching in OpenSearch"
authors:
   - upasagar
   - akjain
   - kkhatua
   - kolchfa
date: 2024-10-11
categories:
  - technical-posts
has_science_table: true
meta_keywords: OpenSearch, tiered caching, query performance, on-heap cache, disk-based cache, request cache, cache hit ratio, OpenSearch Benchmark, performance optimization, caching

> meta_description: Learn how tiered caching in OpenSearch enhances query performance by combining on-heap and disk-based caching. Explore its benefits, limitations, and practical use cases in this detailed blog.

Reviewer: Updated meta: 
_meta_keywords:_ tiered caching, disk-based caching, on-heap caching, OpenSearch caching performance, how tiered caching works 

_meta_description:_ Explore how OpenSearch combines on-heap and disk-based caching to handle larger datasets and improve performance. Learn about the trade-offs of tiered caching, how it works, and future developments.

On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on the node. It offers low-latency data retrieval and thereby providing significant performance gains. However, these advantages come with trade-offs, especially as the cache grows, which may lead to performance challenges.

The size of an on-heap cache size is directly tied to the amount of heap memory available on a node, which is both finite and costly. This limitation creates a challenge when trying to store large datasets or handle numerous queries. When the cache reaches its capacity, older queries must often be evicted to make room for new ones. This frequent eviction can lead to cache churn, negatively impacting performance, as evicted queries may need to be recomputed later.

## Is there a better way?

As discussed, on-heap caches have limitations when handling larger datasets. A more effective caching mechanism is *tiered caching*, which uses multiple cache layers, starting with on-heap caching and extending to a disk-based tier. This approach balances performance and capacity, allowing you to store larger datasets without consuming valuable heap memory.

> In the past, using a disk for caching raised concerns because traditional spinning hard drives were slower. However, advancements in storage technology, like modern SSD and NVMe drives, now deliver much faster performance. Although disk access is still slower than memory, the speed gap has narrowed enough that the performance trade-off is minimal and often outweighed by the advantage of increased storage capacity.

Author: ```suggestion
In the past, using a disk for caching raised concerns because traditional spinning hard drives were slower. However, advancements in storage technology, like modern SSD and NVMe drives, now deliver much faster performance. Although disk access is still slower than memory, the speed gap has narrowed enough that the performance trade-off is minimal and often outweighed by the benefit of increased storage capacity.
```

The workload consisted of a mix of queries, categorized by their shard-level latencies:

* **Expensive**: >150 ms
* **Medium**: 10–150 ms
* **Cheap**: <10 ms

The following diagram presents the performance test results. A red vertical line in the results denotes the baseline percentiles, where the default on-heap cache is enabled and tiered caching is disabled. We tested with 0%, 30%, and 70% query repeatability, corresponding to different cache hit ratios.

![Performance test results](/assets/media/blog-images/2024-10-11-tiered-cache/performance_results.png){:class="img-centered"}

Initial results show that tiered caching performs well, especially with higher cache hit ratios and latencies below p75. The gains are particularly notable in running computationally expensive queries, because tiered caching reduces the need to recompute them, fetching results directly from the cache instead.

## What’s next?

> While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](link). We encourage you to try tiered caching in a non-production environment and share any feedback to help us make this feature more robust.

Reviewer 2: ```suggestion
While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](link). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.
```

> While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](link). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.

Author: ```suggestion
While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](https://github.com/opensearch-project/OpenSearch/issues/10024). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.
```

---
layout: post
title:  "Tiered caching in OpenSearch"
authors:
   - upasagar
   - akjain
   - peteral
   - kkhatua
   - kolchfa

> date: 2024-10-11

Reviewer: Update blog date to 2024-10-24

categories:
  - technical-posts
has_science_table: true
meta_keywords: tiered caching, disk-based caching, on-heap caching, OpenSearch caching performance, how tiered caching works
meta_description: Explore how OpenSearch combines on-heap and disk-based caching to handle larger datasets and improve performance. Learn about the trade-offs of tiered caching, how it works, and future developments.

> ---

Reviewer: Please update the meta with the following: 

meta_keywords:  tiered caching, on-heap cache, disk-based caching, how tiered caching works, OpenSearch cache optimization

meta_description:  Explore the benefits of combining on-heap and disk-based caching in OpenSearch to manage large datasets. Learn how tiered caching works, when to use it, and the performance results of our testing.

For performance-intensive applications like OpenSearch, caching is an essential optimization. Caching stores data so that future requests can be served faster, improving query response times and application performance. OpenSearch uses two main cache types: a request cache and a query cache. Both are on-heap caches, meaning their size is determined by the amount of available heap memory on a node.

## On-heap caching: A good start, but is it enough?

> On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs: as the cache grows, on-heap caching may lead to performance challenges.

Author: ```suggestion
On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs: As the cache grows, on-heap caching may lead to performance challenges.
```

> On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs, especially as the cache grows, which may lead to performance challenges.

Reviewer 3: > However, these advantages come with trade-offs, especially as the cache grows, which may lead to performance challenges.

Maybe it is just me, but this line reads slightly incomplete or looks confusing

Author: Reworded.

Reviewer 4: @kolchfa-aws Can we reword this to something like below? As it is more specific.

"However, these advantages come with trade-offs, especially as the cache grows in size and reaches its capacity, which may lead to performance challenges due to high evictions and misses."

![Tiered cache algorithm](/assets/media/blog-images/2024-10-11-tiered-cache/tc_df_2.png){:class="img-centered"}

Currently, OpenSearch supports the tiered caching model only for the **request cache**. By default, the request cache uses the on-heap cache tier. The cache size is configurable and defaults to 1% of the heap memory on a node. You can enable tiered caching to add a disk-based cache tier, which stores larger datasets that don't fit in memory. This offloads the on-heap cache, improving overall performance.

Tiered caching is also designed to be pluggable. You can seamlessly integrate different types of on-heap and disk cache implementations or libraries using tiered cache settings. For more information, see [Tiered cache](https://opensearch.org/docs/latest/search-plugins/caching/tiered-cache/).

## When to use tiered caching

> Because tiered caching currently only applies to the request cache, it's useful when the existing on-heap request cache isn't large enough to store your datasets and you encounter frequent evictions. You can check request cache statistics using the `GET /_nodes/stats/indices/request_cache` endpoint to monitor evictions, hits, and misses. If you notice frequent evictions along with some hits, enabling tiered caching could provide a significant performance boost.

Reviewer 3: can we roughly quantify `along with some hints`?

Reviewer 4: @kolchfa-aws @jainankitk I think we can leave it as it is. I deliberately left it like that i.e. not quantify it as it is hard to say whether tiered cache will only benefit with >50% or >30% cache hit ratio.

Tiered caching is especially beneficial in these situations:

> - Your domain experiences many cache evictions and has repeatable queries. You can confirm this by using request cache statistics.

Reviewer 3: nit: repeatable -> repeating/repeated ?

> - You're working with log analytics or read-only indexes, in which data doesn't change often, and you're encountering frequent evictions.

Reviewer 3: nit: indexes -> indices

> in which data doesn't change often

looks redundant I guess. read-only indices is self explanatory

Author: Indexes is used per our style guide.

> By default, the request cache only stores aggregation queries. You can enable caching for specific requests by using the `?request_cache=true` query parameter.

Reviewer 3: > stores aggregation queries

stores aggregation query results

The workload consisted of a mix of queries categorized by their shard-level latencies:

* **Expensive**: >150 ms
* **Medium**: 10–150 ms
* **Cheap**: <10 ms

The following diagram presents the performance test results. The red vertical line denotes the baseline percentiles, where the default on-heap cache is enabled and tiered caching is disabled. We tested with 0%, 30%, and 70% query repeatability, corresponding to different cache hit ratios.

![Performance test results](/assets/media/blog-images/2024-10-11-tiered-cache/performance_results.png){:class="img-centered"}

The initial results show that tiered caching performs well, especially with higher cache hit ratios and latencies lower than p75. The gains are particularly notable when running computationally expensive queries because tiered caching reduces the need to recompute them, fetching results directly from the cache instead.

## What's next?

> While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](https://github.com/opensearch-project/OpenSearch/issues/10024). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.

Reviewer 3: > share your feedback to help make this feature more robust

share your feedback to help improve this feature

> While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](https://github.com/opensearch-project/OpenSearch/issues/10024). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.

Reviewer 3: > restarts, or integrating

restarts, and integrating

## On-heap caching: A good start, but is it enough?

> On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs: as the cache grows, on-heap caching may lead to performance challenges.

Reviewer 2: Capitalize "As" after the colon.
